### ISSUE TITLE
Add Display P3 wide gamut colors profile to Flutter (a.k.a. enable 25% more colors in Flutter)

### ISSUE THREAD

**Issue by tomasbaran** [flutter/flutter#55092]
#### 25% more colors
First off, my app is full of imagery. iPhones (since iPhone 7), iPads, Macs and Android devices have been using Display P3 wide gamut for some years now. It's kinda last century to give users the old sRGB colors, when their phones support the wider P3 gamut that provides 25% larger color space. Important especially for images.

#### Color discrepancy btw iOS and Android
Second, I'm not sure it's related to the same issue but there is a discrepancy between colors on Android and iOS when running from the same Flutter codebase. More here:  https://github.com/flutter/flutter/issues/39113 

#### Flutter shall achieve the same quality as the native code
Third, when writing a native iOS app, you can choose which Color Profile you want to write the app in (sRGB/P3). So, when choosing to write a Flutter app (instead of a native one), you are limited and won't achieve the same results. We want to achieve the same in Flutter as in native!

## Proposal

Please, add an option to choose which color Profile we want to write Flutter apps, so they truly are beautiful as they are meant be. It is one of the Flutter pillars, right?

## More info by Apple, Google

- https://developer.android.com/training/wide-color-gamut
- https://developer.apple.com/videos/play/wwdc2017/821/

**Comment by tomasbaran** [comment]
@brianosman I guess, you know more about this topic. Can you shed some light on it? Thanks, Brian.

**Comment by fredlee12345678** [comment]
@tomasbaran @brianegan do you have plan on adding Display P3 to Flutter?

**Comment by tomasbaran** [comment]
@fredlee12345678  I hope @brianegan will be able to tell us more about this topic or even better I hope he or someone else will be able to implement this feature.

I just need that feature in my apps but since I'm just starting with app development—I'm not qualified for implementing it in Flutter.

**Comment by tomasbaran** [comment]
But looking at Apple instructions (https://developer.apple.com/videos/play/wwdc2017/821/) it may not be that difficult to implement:

<img width="1818" alt="Captura de pantalla 2020-05-13 a las 11 30 04" src="https://user-images.githubusercontent.com/25743901/81802148-78155980-950d-11ea-909c-8885c06aaaeb.png">

**Comment by brianegan** [comment]
Hey all -- I seem to have been tagged on this one, but I think I'm the wrong person ^^

**Comment by fredlee12345678** [comment]
It seems none follow this issue?

**Comment by enqvida** [comment]
Wider colors profile should be one of the top priorities. We are in 2022, on devices with screens that are made for rich color presentations. Yet, no matter how great is the design of the app it still looks as soulless as windows 95.

**Comment by hiroshihorie** [comment]
I agree with @enqvida , I thought "building beautiful UI" was a priority for Flutter....

**Comment by Fnalsk** [comment]
I'd also like to add that in general, I feel that optimizing Flutter for usage with mobile devices should take priority over things like windows support. I can't speak for everyone, but I'd imagine that most people using Flutter currently are using it to develop cross platform mobile apps, without particularly caring about desktop. Even if someone wanted to build a desktop app, it would have to have a completely different UI from the corresponding mobile app to deliver a good user experience.

With respect to that this issue in particular being a P4 priority feels very inappropriate. Especially since it's almost 2 years old at this point.

**Comment by ndugger** [comment]
I am personally more invested in flutter from a desktop perspective, but being stuck in legacy color space when P3 support for web is right around the corner is definitely silly. +1 to prioritizing this higher. 

I only just noticed this when I tried to introduce a box shadow and noticed that it's just as choppy if not more choppy than its web counterpart; really makes subtle lighting and shadow effects difficult to implement well.

**Comment by spkersten** [comment]
It appears that on macOS Flutter uses Display P3 (even when it shouldn't). See https://github.com/flutter/flutter/issues/96733#issuecomment-1107830072

**Comment by Krispy145** [comment]
I really need display P3 for iOS soon. Are there any developments with this issue?

**Comment by StefanReinDexcom** [comment]
Here is some information I could gather:

 
### Chromium
Example for P3 (works on a Chrome on a Mac):
https://webkit.org/blog-files/color-gamut/

Chromium has already an implementation for the P3 Color Space:
https://github.com/chromium/chromium/blob/main/ui/gfx/color_space.cc

Also worth a look: 
https://github.com/chromium/chromium/blob/main/ui/gfx/color_space.cc#L696
https://github.com/chromium/chromium/blob/main/ui/gfx/color_space.cc#L915


### Skia
Skia reads and writes the Color Space in the header.

SkImageInfo:
https://github.com/google/skia/blob/main/src/core/SkImageInfo.cpp

contains the SkColorSpace:

https://github.com/google/skia/blob/main/src/core/SkColorSpace.cpp
https://github.com/google/skia/blob/main/src/core/SkICC.cpp


### Flutter
Unsure where Flutter finally draws, but these here look suspicious to me:

https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/ios_surface_software.mm#L62
https://github.com/flutter/engine/blob/main/shell/common/rasterizer.cc#L286
https://github.com/flutter/engine/blob/main/flow/layers/offscreen_surface.cc#L19

If we would have somewhere an attribute we can pass through from Flutter, we could exchange the calls here:

`SkColorSpace::MakeSRGB()`

with this here (or something configurable actually):

`SkColorSpace::MakeRGB(SkNamedTransferFn::kLinear,  SkNamedGamut::kDisplayP3)`


If I got all this wrong, please let us know what needs to be done or how to approach this.

Thanks,
Stefan

**Comment by serena7889** [comment]
Our development team has transitioned to flutter because of the speed it allows us to develop for all the platforms we're trying to reach, but with design being such an important part of out product, not being able to use DisplayP3 colors (particularly on iOS) might be a deal breaker for us. 

I'd like to request the priority of this gets increased due to how drastic the difference between Flutter apps (with sRGB) and native apps (with DisplayP3). Until this is sorted, for us and many other teams, Flutter is not a viable alternative.

**Comment by StefanReinDexcom** [comment]
@serena7889 I've seen that you forked the Flutter Engine.

Could you get it working? I would be open to invest my time into this and help.
Please let me know, because chances are people will try in parallel, like me (I originally would start in a couple of weeks and try myself).
I've followed you and put my email into my bio on GitHub - don't know how I otherwise could email you myself about this.

My naive approach would be first that I change all the occurrences I mentioned above with `SkNamedGamut::kDisplayP3` and then compile the code and see what happens. 


Thanks,
Stefan

**Comment by armandojimenez** [comment]
Any plans for this?

**Comment by enqvida** [comment]
@Elysium-Planitia What exactly needs to be fixed?

**Comment by Elysium-Planitia** [comment]
@enqvida It would be great if there was an option to set the color space to the one the display has. If there'd be a Flutter native way to switch the color spaces then things like displaying images and other graphics from a user would be possible. As of right now, there is no way for a Flutter application to display images in a correct way all the time. Editing is also impossible and so are the potential apps built with Flutter. This is true not only for Desktop but also for mobile devices. You could argue that it's similar to the maximum 60fps of Flutter on higher frequency screens. But missing colors is something that makes correct visual representations in the above mentioned cases impossible instead of just not as smooth as other apps might be like it is with the framerate.

**Comment by Elysium-Planitia** [comment]
I'd like to add, that the current slogan on flutter.dev is "Build apps for any screen". I want to do  but I can't.

**Comment by jabedamin** [comment]
Unfortunately most serious tech company will immediately rule flutter out from its tech stack until this is supported. Almost all smartphone displays nowadays support wide gamut color. Has anyone from Flutter even acknowledged this issue yet? Are there any plans in the roadmap to address it? @timsneath

**Comment by armandojimenez** [comment]
I don't understand how this is not a bigger issue. I don't understand how people don't notice that colors in Flutter looks washed out and lifeless. Can someone from the Flutter team please acknowledge this issue?

**Comment by Elysium-Planitia** [comment]
more than five years of display P3 with Apple (wwdc2017) and not even an acknowledgement of Flutter that this is an issue only getting more and more serious? I'm done with Flutter. I'm learning Swift/SwiftUI now. If there is a possibility to render Images in P3 while making all computations in Dart/Flutter that would be at least be an option. But as far as I know not even this is possible? Bye Flutter...

**Comment by jtkeyva** [comment]
does impeller address/solve this?

**Comment by jmagman** [comment]
Capturing discussion around this:

> We tried color correct rendering in the past but folks didn't like SRGB colors converted to wider gamuts and the colors looking washed out [SRGB colors stretched into a wider gamut without additional conversion]. Android manufacturers choose to boost colors https://github.com/flutter/flutter/issues/39113#issuecomment-525766991. We backed out our changes in response. Also, we need a way for callers to specify the colorspace.

This would require work an API to be exposed and used by the framework to specify/grab the color profiles. Whatever pattern works with Skia would also likely work with Impeller, but would need to be implemented either way.

**Comment by jtkeyva** [comment]
There is a huge demand by professionals to support the greatest quality a device can handle. Can you re-enable and give developers a switch or declare what color space to use?

**Comment by jmagman** [comment]
Yes, it's possible this would be opt-in feature (possibly via an Info.plist key) to be plumbed through.

**Comment by jtkeyva** [comment]
@jmagman great news! Soon enough we'll be able to silence any disputes about Flutter looking dull or flat and not ready for prime time :)

**Comment by ajGulati05** [comment]
@jmagman I was wondering if there is a timeline on this? Thank you

**Comment by jmagman** [comment]
No timeline to report.

**Comment by CoderUni** [comment]
Could you raise the priority of this issue @jmagman @stuartmorgan ? Having a consistent android and ios UI is a must. Having dull colors also makes it less appealing for ios users. Thanks!

**Comment by jmagman** [comment]
P4 is default for bugs that are not currently being worked on, we are aware of its importance.  Please see https://github.com/flutter/flutter/wiki/Issue-hygiene#when-will-my-bug-be-fixed for prioritization policies.

**Comment by CoderUni** [comment]
Thanks @jmagman , I hope this issue is something that the Flutter team would work on soon.

**Comment by jtkeyva** [comment]
Hopefully this is in the P0-P3 range soon! Interestingly the feature itself is called P3 haha.

Sometimes the way something looks is more important than the way it functions underneath to some people. This goes for non-techies and people who want to LOOK as good as a native app. All of us iOS Flutter devs are driving around nice reliable cars with faded paint jobs. We want the shine and polish too! Design and aesthetic is a massive priority to the outside world.

**Comment by jtkeyva** [comment]
hey @serena7889 and @StefanReinDexcom are you making any progress on this? i don't have the low level experience but maybe i can put a tiny bit of money behind it to help expedite?

**Comment by pvaibhav** [comment]
I was considering using Flutter to develop a colour calibration tool (for internal use) but had to give up on the idea because it can't support P3 gamut. It was not critical for the app since its primary purpose was to _measure_ and not display -- still the inability to show any colours outside the sRGB/Rec.709 space is, to be frank, inexplicable in 2022 for a UI framework. Laptops and phones with DCI P3 gamut have been the norm for 6-7 years now. Please consider prioritising this issue. Thanks.

**Comment by jtkeyva** [comment]
https://www.reddit.com/r/FlutterDev/comments/yywx66/attn_all_your_vote_is_needed_to_fix_critical/

**Comment by jmagman** [comment]
Please just 👍 the issue to subscribe to changes, do not add "me too" comments--it adds noise to the issue tracker and makes it harder to find the relevant information to actually fix this issue.  We know it's important, that's why it's an open issue and hasn't been closed.  We actually are planning on working on it.  🙂

**Comment by rohiti** [comment]
Can someone please help me understand this issue in very layman terms? I am facing something very similar and I did an extensive research on my end, but haven't been able to figure out what exactly happens when the app runs on iOS?

Is it something like -> I give a Color code in flutter (0xFF282726) -> iOS converts this to P3 into (R: 0.1569, G: 01529, B: 0.1490) -> (Skia maybe?) Converts it back to sRGB (R: 0.1578, G: 0.1528, B: 0.1485) and this gets displayed on the screen which is a lot more dull than the original color?


Apologies for asking such a naive question on this thread, but we had spent more than 3 months in developing the first version of our app in flutter and after coming across this issue, we have to make the painful decision of writing the entire code in native for both iOS and Android for another three months. Please help me understand this, it would be great if I can apply some workaround to this to get the first version rolled out.

**Comment by gaaclarke** [comment]
@tomasbaran I just want to set expectations since the description of this issue says "25% more colors" and "Important especially for images."  Supporting rendering in a wide gamut colorspace will not give you more colors from images since the file formats supported by Flutter are limited to 8bits per channel.  For images in the Display P3 colorspace, you'll just get different colors (correct colors and a higher dynamic range) on the display.  For images in the sRGB colorspace they will look the same.

What are the images you are trying to render?  Are they jpegs in the Display P3 colorspace?  What about other colorspaces like Adobe RGB?

> Please, add an option to choose which color Profile we want to write Flutter apps, so they truly are beautiful as they are meant be. It is one of the Flutter pillars, right?

Why are you considering an option to add support instead of just doing it all the time?

**Comment by knopp** [comment]
> Why are you considering an option to add support instead of just doing it all the time?

Supporting wide gamut comes with an overhead. The backing stores (surface + raster cache) would need wide gamut pixel format (i.e. bgra10_xr_srgb onAapple, which is 64bit per pixel), or separate 32 bit surface for color (10 bit per channel) + 8 bit surface for alpha, which I'm not sure skia supports.

So if application doesn't have any wide gamut contents it might be good to be able to opt-out this. Also in case when application is running on device that doesn't support wide-gamut.

**Comment by StefanReinDexcom** [comment]
@knopp Skia does support P3. It works on Chrome, which uses Skia. Chromium has an implementation for this.

Somewhere up I even showed the LOC where to start from my point of view.

**Comment by knopp** [comment]
I never questioned whether Skia supports P3, my question was about support separate buffer for color attachment and alpha channel, which would allow doing wide-gamut with 40bits per pixel instead of 64.

**Comment by gaaclarke** [comment]
Thanks @knopp, that's a good assessment.  I was considering the memory requirements and opt-in in my design doc.  I was more fishing for the reporters' expectations.  I think there are some misconceptions about what adding Display P3 will do =)

**Comment by gaaclarke** [comment]
I've written up a design doc that proposes how to implement this feature in the engine: https://docs.google.com/document/d/17C7UAaLSX2hHOq-hYENeJ1HwMhPCL-e0pb9Pn8rXUok/edit

TLDR: Expand the bitrate for render buffers to 64bits / pixel and keep the color encoding in linear space.  Add support for Display P3 images and smoother gradients, avoid any Dart API changes.

**Comment by leighajarett** [comment]
Hi all! We added a new flag for wide gamut support on master. It would be great to get some testers! You can enable it with [this flag](https://github.com/flutter/flutter/blob/a463bb82c413af319c8b715deedd7f2c345becd6/dev/integration_tests/wide_gamut_test/ios/Runner/Info.plist#L52). Make sure you're using Impeller and that you're running the app on a real device, not a simulator. Please send me a message on discord (@leighajarett) if you're giving it a try :)

**Comment by jabedamin** [comment]
> Hi all! We added a new flag for wide gamut support on master. It would be great to get some testers! You can enable it with [this flag](https://github.com/flutter/flutter/blob/a463bb82c413af319c8b715deedd7f2c345becd6/dev/integration_tests/wide_gamut_test/ios/Runner/Info.plist#L52). Make sure you're using Impeller and that you're running the app on a real device, not a simulator. Please send me a message on discord (@leighajarett) if you're giving it a try :)

Great news! I'd like to test it, what version of flutter is it available from?

**Comment by jmagman** [comment]
https://github.com/flutter/engine/pull/39111 behind the `FLTEnableWideGamut` Info.plist value is available in Flutter 3.8.0-12.0.pre, so not quite yet in the [beta channel](https://docs.flutter.dev/development/tools/sdk/releases?tab=macos).  You'll need to try it on the [master channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels) for now.

**Comment by jabedamin** [comment]
> [flutter/engine#39111](https://github.com/flutter/engine/pull/39111) behind the `FLTEnableWideGamut` Info.plist value is available in Flutter 3.8.0-12.0.pre, so not quite yet in the [beta channel](https://docs.flutter.dev/development/tools/sdk/releases?tab=macos). You'll need to try it on the [master channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels) for now.

Thanks, that's useful. Of course is only an iOS setting, does this release also have support for android devices?

**Comment by jmagman** [comment]
> does this release also have support for android devices?

No, we started with iOS.  From https://docs.google.com/document/d/17C7UAaLSX2hHOq-hYENeJ1HwMhPCL-e0pb9Pn8rXUok/edit#heading=h.t6z7bji7vfnm

> I propose we should prioritize iOS since it represents our largest user base with access to wide gamut displays.  Other platforms can be considered later.  However, I did do some research above to convince the team that we aren't painting ourselves into the proverbial corner with respect to Android.

**Comment by jmagman** [comment]
> [flutter/engine#39111](https://github.com/flutter/engine/pull/39111) behind the `FLTEnableWideGamut` Info.plist value on iOS only is available in Flutter 3.8.0-12.0.pre

This is now in the [beta channel](https://docs.flutter.dev/development/tools/sdk/releases?tab=macos).

**Comment by fiercepark** [comment]
> > [flutter/engine#39111](https://github.com/flutter/engine/pull/39111) behind the `FLTEnableWideGamut` Info.plist value on iOS only is available in Flutter 3.8.0-12.0.pre
> 
> This is now in the [beta channel](https://docs.flutter.dev/development/tools/sdk/releases?tab=macos).

![21684140849_ pic copy](https://github.com/flutter/flutter/assets/86466287/73f7c285-0ae1-4b01-9e7a-66368b861c65)
We have a great p3 colour space in flutter 3.10, a great milestone, but Color(0xff${hex}) is currently displayed incorrectly.

**Comment by gaaclarke** [comment]
@fiercepark can you file an issue if you've identified a bug, please.

**Comment by gaaclarke** [comment]
@jabedamin wide gamut colors are only displayed if you have wide gamut images or are using gradients.  Please file an issue with a screenshot if you expect wide gamut colors and aren't seeing it.

**Comment by jabedamin** [comment]
@gaaclarke I see, I was hoping flutter objects would also support wide gamut colours in addition to images. For example Button color or Container(color: Colors.blueAccent), is that not supported? If not, any plans to support in the future? It's these colours that are sometimes more visibly dull than images.

**Comment by gaaclarke** [comment]
@jabedamin we're never going to change the existing code to mean something different.  We might add new colors in the future, something like `Container(color: Color.displayP3(0.0, 0.0, 1.0, 1.0))`.

**Comment by jabedamin** [comment]
@gaaclarke yes that would help hugely. I suspect many others are interested in this too, in order for Flutter UIs to level up to native apps on Swift.

This might be my lack of understanding on how the code behind the scenes renders colours in Flutter. I was thinking more in the sense of displaying colours based on available colour spaces on any given device. So when designing an app, if the designer uses for example RGB(0, 209, 169) on an object - this RGB code would not not change, but it would be displayed differently depending on the device and supported colour space.

**Comment by jtkeyva** [comment]
@gaaclarke and @jabedamin yeah this almost feels like a "gotcha" moment that was too good to be true. of course i appreciate the p3 support for images, i fought for it! i assumed from the title "Add Display P3 wide gamut colors profile to Flutter (a.k.a. enable 25% more colors in Flutter)" it would be 25% more colors across the entire app ui. seems a shame to have vibrant images and still a flat dull UI if it's capable.

the end user is the customer and the customer likes the bright shiny silky smooth end product. what can we do to push this along the priority?

**Comment by gaaclarke** [comment]
Okay, here's what we are going to do.  This issue is very broad and I believe that the majority of what it represents has now been fulfilled on `main`.  Namely:

1) iOS Display P3 support
1) Display wide gamut images correctly
1) Take advantage of higher bits/pixel for gradients

All of that is on `main` and on by default now.  So, I'm going to close this issue and open more specific issues that represent any remaining potential work that's been discussed here so people can discuss them / upvote them:

1) Add support for Android -- https://github.com/flutter/flutter/issues/127852
1) Add support for the Framework -- https://github.com/flutter/flutter/issues/127855

If there is anything else I missed, please file an issue (including if you have found a bug in wide gamut image support).

**Comment by jtkeyva** [comment]
thank you!

for those who aren't following this closely, is it ok to post this in reddit r/flutterdev to inform people? just as i thought it was a huge win, i don't think the community there knows that it's only for wide gamut images.

is it ok to notify them ell them where to go to and discuss / upvote it if they find it important to them to include in the framework?

i don't want to get dinged for "farming" upvotes like someone warned about
